deepspeed/profiling/flops_profiler/README.md

```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
-- `end_profile` - ends profiling and cleans up, invoked at the end of the profiling and before any printing method.
-
-`flops_to_string`, `params_to_string`, `duration_to_string` are utility functions to convert the metric number to string.
-
-Below is an example of this usage in a typical training workflow.
+The flops profiler can be enabled by adding the following field to the `deepspeed_config` file.
 
-```python
-from deepspeed.profiling.flops_profiler.profiler import FlopsProfiler
-
-model = Model()
-profiler = FlopsProfiler(model)
-
-start_step = 5
-end_step = 10
-assert (end_step > start_step), "should end profiling after start profiling"
-print_profile = True
-pring_aggregated_profile = True
+```json
+{
+  "flops_profiler": {
+    "enabled": false,
+    "profile_step": 1,
+    "module_depth": -1,
+    "top_modules": 3
+  }
+}
+```
 
-for step, batch in enumerate(data_loader):
-  # start profiling at training step "profile_step"
-  if step == start_step:
-    profiler.start_profile()
-
-  # end profiling and print output at training step "profile_step"
-  if model == end_step: # if using multi nodes, check global_rank == 0 as well
-    flops = profiler.get_total_flops()
-    params = profiler.get_total_flops()
-    duration = profiler.get_total_duration()
-    steps = profiler.get_total_steps()
-    if print_profile:
-        profiler.print_model_profile()
-    if print_aggregated_profile:
-        profiler.print_model_aggregated_profile(module_depth=-1, top_modules=3)
-    profiler.end_profile()
-    print(flops, params, duration, step)
+An example output of 4-layer Megatron-LM model (`hidden_size = 512, num_attention_heads = 16, batch_size = 8, seq_length = 1024`) is shown below.
+
+```shell
+DistributedDataParallel(
+  38.89 M, 100.00% Params, 103.62 GMACs, 100.00% MACs, 33.31 ms, 100.00% time, 6.2 TFLOPS,
+  (module): FP16_Module(
+    38.89 M, 100.00% Params, 103.62 GMACs, 100.00% MACs, 33.29 ms, 99.94% time, 6.2 TFLOPS,
+    (module): GPT2Model(
+      38.89 M, 100.00% Params, 103.62 GMACs, 100.00% MACs, 33.25 ms, 99.80% time, 6.2 TFLOPS,
+      (language_model): TransformerLanguageModel(
+        38.89 M, 100.00% Params, 103.62 GMACs, 100.00% MACs, 4.73 ms, 14.21% time, 4.4e+01 TFLOPS,
+        (embedding): Embedding(
+          26.28 M, 67.57% Params, 0 MACs, 0.00% MACs, 307.56 us, 0.92% time, 0.0 TFLOPS,
+          (word_embeddings): VocabParallelEmbedding(25.76 M, 66.23% Params, 0 MACs, 0.00% MACs, 130.41 us, 0.39% time, 0.0 TFLOPS, )
+          (position_embeddings): Embedding(524.29 k, 1.35% Params, 0 MACs, 0.00% MACs, 68.19 us, 0.20% time, 0.0 TFLOPS, 1024, 512)
+          (embedding_dropout): Dropout(0, 0.00% Params, 0 MACs, 0.00% MACs, 56.27 us, 0.17% time, 0.0 TFLOPS, p=0.1, inplace=False)
+        )
+        (transformer): ParallelTransformer(
+          12.61 M, 32.43% Params, 103.62 GMACs, 100.00% MACs, 4.4 ms, 13.22% time, 4.7e+01 TFLOPS,
+          (layers): ModuleList(
```

Comment:
Why is time percent and TFlops 0 for ModuleList?

Reply:
Fixed.